language/compiler/ir-to-bytecode/src/compiler.rs

```diff
@@ -661,6 +664,32 @@ fn compile_explicit_dependency_declarations(
     Ok(())
 }
 
+fn compile_friends(
+    context: &mut Context,
+    address_opt: Option<AccountAddress>,
+    friends: Vec<ast::FriendDeclaration>,
+) -> Result<Vec<ModuleHandle>> {
+    let mut friend_decls = vec![];
+    let mut friend_decls_set = HashSet::new();
+    for friend in friends {
+        let ident = match (address_opt, friend.ident) {
+            (Some(address), ModuleIdent::Transaction(name)) => {
+                QualifiedModuleIdent { address, name }
+            }
+            (None, ModuleIdent::Transaction(name)) => bail!(
+                "Invalid friend '{}'. No address specified for script so cannot resolve friend",
+                name
+            ),
+            (_, ModuleIdent::Qualified(id)) => id,
+        };
+        let handle = context.declare_friend(ident)?;
+        if friend_decls_set.insert(handle.clone()) {
+            friend_decls.push(handle);
+        }
```

Comment:
I'm not convinced this check should be here. It would make it easier to test duplicate friend decls

Reply:
Right! It should not be here. I'll have it removed. The duplication checker already performs the check.

Reply:
+1. Writing the IR compiler is always fun/counterintuitive...

Reply:
This one is sort of a fun one, because we generally don't get to make malformed pools in the IR (since it is auto generated for the most part)

So this one is sort of neat. I'm not sure if we do this sort of eager population for imports too, but we should make sure we do

Reply:
The check is removed, and this allows us to write a new test case, `friend_decl_duplicated.mvir`, which is included in this PR as well.